### PR TITLE
feat(math): add ideal membership and normal form operations (#1665)

### LINKED ISSUE #1665: [Feature]: Add bounded exact ideal membership and quotient ring reduction
## Summary

Jacobian has polynomial Gröbner basis computation (`polynomial.groebner_basis.compute`) and Jacobian syzygy operations, but no operation for ideal membership testing or quotient ring reduction. The conjecture sources (Jacobian conjecture counterexamples, finite-field Jacobian syzygies over quotient rings in #964, Koszul filtration conjectures) need exact ideal membership and quotient ring arithmetic.

## Current behavior

- `polynomial.groebner_basis.compute` computes a Gröbner basis for an ideal.
- `polynomial.jacobian_syzygy.*` computes Jacobian syzygies over `QQ`.
- There is no operation that tests whether one polynomial belongs to a declared ideal, or reduces one polynomial modulo an ideal to its canonical normal form in the quotient ring.
- Issue #964 tracks finite-field Jacobian syzygies over quotient rings but not general ideal membership or quotient ring reduction.

## Requested capability

Add bounded exact ideal membership and quotient ring reduction operations:

- `polynomial.ideal.membership.decide`: given an ideal (generators + coefficient field + monomial order) and one polynomial, decide membership and return the remainder (normal form) if not in the ideal, or the explicit representation if in the ideal.
- `polynomial.quotient_ring.reduce`: given an ideal and one polynomial, reduce to canonical normal form in the quotient ring `R[x_1,...,x_n]/I`.

Each operation:
- accepts generators, coefficient field, ordered variables, monomial order, and one polynomial;
- is backed by SymPy's Gröbner basis reduction or Singular — not hand-rolled reduction;
- fails closed on unsupported fields, timeout, or malformed ideals;
- returns `COMPUTED` evidence.

## Why this is in scope

Jacobian's AGENTS.md states that operations may use SymPy or Singular as a private computational engine. Ideal membership and quotient ring reduction are atomic operations that return concrete mathematical values (a decision and a normal form), not workflows or durable state. These compose directly with the existing Gröbner basis operation.

## Acceptance criteria

- [ ] `polynomial.ideal.membership.decide` is supported with a producer.
- [ ] `polynomial.quotient_ring.reduce` is supported with a producer.
- [ ] Tests cover a polynomial in the ideal (remainder = 0), a polynomial not in the ideal (nonzero remainder), and a quotient ring reduction over `QQ`.
- [ ] The operations fail closed on unsupported fields and malformed ideals.
- [ ] The operations are discoverable via `math.find` and executable via `math.run`.

## Related

- Existing: `polynomial.groebner_basis.compute`, `polynomial.jacobian_syzygy.*`
- #964 — finite-field Jacobian syzygies over quotient rings
- Conjecture sources: Jacobian conjecture counterexamples, Koszul filtration conjectures

### PR DESCRIPTION
**Pull request by morluto**
## Summary

Add two atomic operations for exact polynomial ideal computations over QQ:

- `polynomial.ideal.membership.decide`: decides whether a bounded rational polynomial lies in a bounded ideal, returning both the decision and the canonical normal form (remainder)
- `polynomial.ideal.normal_form.compute`: reduces a polynomial modulo an ideal using a Gröbner basis under an explicit monomial order

## Design

Both operations use SymPy's `groebner` basis computation followed by `basis.reduce()` as the computational backend. The operations accept bounded rational polynomial generators with an explicit monomial order (lex, grlex, or grevlex) and use the existing `RationalPolynomial` canonical type.

This follows the bitter lesson: ideal membership and normal form reduction are fundamental algebraic primitives that compose across algebraic geometry, invariant theory, and computational algebra — they are not task-specific solutions.

## Tests

- Member and non-member detection (univariate)
- Zero in any ideal
- Canonical remainder computation
- Mismatched ring rejection
- Multivariate membership and non-membership

Closes #1665

[Continue this on Linzumi](https://serve.linzumi.com/w/williamlin1327-workspace/thread/general/636bb865-8fa9-4348-a437-f0b719b854e9)

### CODE REVIEW

File: src/jacobian/math/polynomials/_models.py [hunk 1]
```diff
@@ -316,6 +320,83 @@ def require_canonical_basis_ring(self) -> Self:
...
+                gen,
+                maximum_terms=MAX_POLYNOMIAL_TERMS,
+                maximum_exponent=12,
+                maximum_coefficient_digits=128,
+                label="ideal generator",
+            )
+        # The normal form of the zero ideal is the input polynomial itself;
+        # cap the input at the 1,024-term result boundary so an accepted
+        # request can never leak a result-budget host exception.
+        if len(self.polynomial.polynomial.terms) > _MAX_RESULT_POLYNOMIAL_TERMS:
```
[the remark this answers is not in the document]
Comment (morluto) [reply src/jacobian/math/polynomials/_models.py]: Fixed in 31d25e0f4: both catalog examples now nest `variables`/`generators` under the canonical `ideal` field; the example-conformance integration test executes them successfully.

File: src/jacobian/math/polynomials/_models.py [hunk 2]
```diff
@@ -316,6 +320,83 @@ def require_canonical_basis_ring(self) -> Self:
...
+    """
+
+    in_ideal: bool
+    normal_form: RationalPolynomial
+
+    @model_validator(mode="after")
+    def bind_membership_to_normal_form(self) -> Self:
+        remainder_is_zero = not self.normal_form.polynomial.terms
+        if self.in_ideal is not remainder_is_zero:
+            raise ValueError("in_ideal must equal (normal_form == 0)")
```
[the remark this answers is not in the document]
Comment (morluto) [reply src/jacobian/math/polynomials/_models.py]: Fixed in 31d25e0f4: both results now retain the source ideal, source polynomial, monomial order, and computed Gröbner basis. Validation replays the defining reduction relation (unique remainder modulo the retained basis plus generator absorption) and rejects detached or tampered payloads.

File: src/jacobian/math/polynomials/_models.py [hunk 3]
```diff
@@ -316,6 +320,83 @@ def require_canonical_basis_ring(self) -> Self:
...
+    serialized ideals compose without unpacking.  The request polynomial is
+    bounded to 1,024 terms because the zero ideal returns it unchanged as
+    the normal form and the exact-result contract caps at 1,024 terms.
+    """
+
+    ideal: RationalPolynomialIdeal = Field(
+        description=(
+            "A bounded polynomial ideal in one ordered QQ ring: at most 16 "
+            "generators with at most 256 aggregate terms, degree at most 12, "
+            "and coefficient components at most 128 digits."
```
[the remark this answers is not in the document]
Comment (morluto) [reply src/jacobian/math/polynomials/_models.py]: Fixed in 31d25e0f4: admission now enforces at most 16 generators and total degree at most 12 per generator term (subsuming each variable's exponent), matching the advertised backend-work domain.

File: src/jacobian/math/polynomials/_models.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (morluto) [reply src/jacobian/math/polynomials/_models.py]: Fixed in 31d25e0f4: remainder expansion past the exact-result boundary is now a typed outcome. Both results carry status COMPUTED | BUDGET_EXCEEDED (membership: IN_IDEAL | NOT_IN_IDEAL | BUDGET_EXCEEDED); when the reduced form exceeds the 1,024-term boundary the operation returns BUDGET_EXCEEDED with no partial remainder instead of raising PolynomialOutputBudgetError. The reproduced counterexample (<x-(1+y+z)^12>, x^12, lex) now returns the typed outcome; a regression test covers it plus an in-budget-boundary case (<x-(1+y+z)^4>, x^12).

File: src/jacobian/math/polynomials/_replay.py [hunk 4]
```diff
@@ -0,0 +1,352 @@
...
+        try:
+            basis = groebner(current, *symbols, order=monomial_order, domain=QQ)
+        except Exception:
+            return None, False
+        try:
+            exceeded = _basis_exceeds_output_budget(basis, symbols)
+        except Exception:
+            return None, False
+        if exceeded:
+            return None, True
```
[the remark this answers is not in the document]
Comment (morluto) [reply src/jacobian/math/polynomials/_replay.py]: Fixed in 2e79ff996: the output-basis decision is deferred to the complete source basis. The incremental kernel now checks intermediate prefix bases only against a separate 4,096-term work envelope (plus exponent/coefficient representability), while the public 1,024-term output boundary is applied solely to the final basis after every generator is incorporated — so an oversized prefix can never decide the outcome, with or without a trailing constant. Generator processing order is now derived canonically from mathematical content (ascending total degree, then term count, then term fingerprint), making the kernel a deterministic function of the ideal value and removing presentation-order dependence entirely (subsuming the earlier constant-generator special case). Regression test test_later_generator_collapse_not_reported_as_budget_overflow covers your <x-(1+y+z)^4, x^12, 1> example; test_budget_outcome_is_independent_of_generator_presentation_order asserts identical outcomes across all generator permutations.

File: src/jacobian/math/polynomials/_replay.py [hunk 5]
```diff
@@ -0,0 +1,367 @@
...
+    symbols = symbols_for_variables(variables)
+    try:
+        exprs = [rational_polynomial_to_sympy(gen).as_expr() for gen in ideal.generators]
+    except Exception:
+        return None, False
+    # A nonzero constant generator collapses the ideal to the unit ideal no
+    # matter how an oversized prefix basis looks; presentation order must
+    # not turn that collapse into a false budget overflow.
+    for expr in exprs:
+        if getattr(expr, "is_Number", False) and expr != 0:
```
[the remark this answers is not in the document]
Comment (morluto) [reply src/jacobian/math/polynomials/_replay.py]: Fixed in 2e79ff996. Reproduced: your lex request <x+f, x^12, x> (f = all 15 monomials of degree at most 4 in y,z) returned BUDGET_EXCEEDED while the same ideal presented as <x, x^12, x+f> returned IN_IDEAL — the two-generator prefix basis {x+f, f^12} (1,241 aggregate terms) was tripping the public 1,024-term output boundary before the trailing x was ever read. The fix separates the budgets: prefix bases are checked only against a declared 4,096-term intermediate work envelope (prefixes may legitimately exceed the output boundary and still shrink when a later generator grows the ideal), and the 1,024-term boundary is decided exclusively on the complete final basis — here <x, f>, 16 terms, so both operations now return COMPUTED / IN_IDEAL with normal form zero. Processing order is canonical (content-derived: ascending total degree, then term count, then term fingerprint), so x sorts first anyway and the outcome is provably independent of presentation order; regression tests cover your exact example, all six permutations asserting identical status, and the negative case (<x-(1+y+z)^4, x^12>, whose complete 1,225-term reduced basis genuinely exceeds the boundary) still reporting BUDGET_EXCEEDED.

File: src/jacobian/math/polynomials/_replay.py [hunk 6]
```diff
@@ -0,0 +1,415 @@
...
+            basis = groebner(current, *symbols, order=monomial_order, domain=QQ)
+        except Exception:
+            return None, False
+        try:
+            if _basis_exceeds_output_budget(
+                basis, symbols, enforce_aggregate_terms=False
+            ):
+                # Representability is scale-free: this exact artifact can
+                # never be materialized as canonical values.
+                return None, True
```
[the remark this answers is not in the document]
Comment (morluto) [reply src/jacobian/math/polynomials/_replay.py]: Root cause: `incremental_source_groebner` treated a representability trip (exponent/coefficient limits) at an intermediate prefix as final evidence, but adding generators grows the ideal yet can shrink its reduced basis — your lex chain's five-generator prefix carries x5^161051 while the complete ideal is the unit ideal, so the reported BUDGET_EXCEEDED was false.

Fix (src/jacobian/math/polynomials/_replay.py, _operations.py, _models.py): no prefix ever decides an output budget now.
- Guarded strategy 1 (canonical ascending order) and strategy 2 (reverse order) abort only their own work bound at a mid-sequence trip; only the basis of the complete generating set decides overflow.
- Strategy 3 runs one unguarded kernel call over the complete generating set in a killable subprocess with a 10 s wall clock (`run_bounded_process`, the repo's subprocess-confined boundary), where Buchberger sees every collapsing pair from the start — your counterexample completes instantly as basis (1).
- When no bounded strategy concludes (e.g. the eight-variable C-chain whose true basis has ~34M-digit coefficients), both operations report a new typed UNKNOWN outcome that asserts nothing mathematical and carries no partial artifact; validators accept it only fully stripped.

Validation: new regression test `test_prefix_exponent_overflow_not_reported_for_late_unit_collapse` reproduces your exact ideal and asserts COMPUTED/IN_IDEAL with basis (1); `test_complete_basis_exponent_overflow_reports_typed_outcome` keeps genuine overflow typed BUDGET_EXCEEDED; the C-chain test now asserts UNKNOWN. ruff check/format clean repo-wide, mypy strict clean, tests/math/polynomials 97 passed, tests/catalog + tests/dispatch + tests/tooling + tests/mcp all green. Pushed as 8b60a8756.

File: src/jacobian/math/polynomials/_invariants.py [hunk 7]
```diff
@@ -244,6 +249,109 @@
...
+        IdealMembershipRequest,
+        IdealMembershipResult,
+        polynomial_ideal_membership,
+        "ideal-membership",
+        "normal-form",
+        "exact",
+        examples=(
+            example(
+                "membership_x_squared",
+                "Decide if x lies in <x^2> in Q[x].",
```
[the remark this answers is not in the document]
Comment (morluto) [reply src/jacobian/math/polynomials/_invariants.py]: Root cause: the two invocation examples violated the repository's two-part example contract — `membership_x_squared` stated only the computation and `normal_form_x_mod_x_squared` stated an expected result instead of an input precondition, leaving the shared-ring requirement and operation-specific bounds discoverable only through validator rejections.

Fix (src/jacobian/math/polynomials/_invariants.py): both descriptions now carry a precondition clause stating that the ideal and polynomial must share one identical ordered QQ ring, with at most 16 generators of per-term total degree at most 12, at most 1,024 queried polynomial terms, and coefficient components of at most 128 digits; the result statement was dropped from `normal_form_x_mod_x_squared`.

Validation: ruff check/format clean, mypy strict clean, tests/catalog and tests/math/polynomials all pass (55 passed). Pushed as 26cf9cd1e.

File: src/jacobian/math/polynomials/_replay.py [hunk 8]
```diff
@@ -0,0 +1,579 @@
...
+    """
+
+    from sympy import QQ, groebner
+
+    current: list[Any] = []
+    basis = None
+    for position, expr in enumerate(exprs):
+        current.append(expr)
+        try:
+            basis = groebner(current, *symbols, order=monomial_order, domain=QQ)
```
[the remark this answers is not in the document]
Comment (morluto) [reply src/jacobian/math/polynomials/_replay.py]: Root cause: agreed — only the third strategy ran behind the killable subprocess; the two guarded incremental strategies called `groebner` in the service process, and the per-prefix budget check could only run after SymPy had fully materialized each prefix, so an admitted hard prefix (e.g. a lexicographic cyclic system within the eight-variable degree-12 limits) could consume unbounded service time or memory first.

Fix (src/jacobian/math/polynomials/_groebner_worker.py, _replay.py): every kernel attempt now runs behind the bounded process boundary. The worker performs the incremental construction itself — one generator per step with a budget check after each prefix under your chosen strategy order — and reports tagged outcomes ("ok" wire basis / "exceeded" on complete-basis overflow / "aborted" for a pure work bound / "failed"); `incremental_source_groebner` is now a pure orchestrator of up to three bounded attempts (ascending, descending, complete). A prefix trip remains a work bound and never decides an outcome; only a complete basis does.

Validation: ruff check/format clean repo-wide, mypy strict clean, `make architecture` gateway OK, lint-imports 4 kept 0 broken, tests/math/polynomials 97 passed. Pushed as 576fe99ec.

File: src/jacobian/math/polynomials/_groebner_worker.py [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+            ]
+            for generator in ideal.generators
+        ],
+    }
+    try:
+        completed = run_bounded_process(
+            [sys.executable, "-c", _WORKER_PROGRAM],
+            input_bytes=json.dumps(payload).encode("utf-8"),
+            timeout_seconds=_TIMEOUT_SECONDS,
+            environment=worker_environment(),
```
[the remark this answers is not in the document]
Comment (morluto) [reply src/jacobian/math/polynomials/_groebner_worker.py:365]: Root cause: agreed — the worker call relied on the 10 s wall clock alone, so an admitted high-coefficient system could allocate enormous integers and exhaust host memory before the deadline.

Fix (src/jacobian/math/polynomials/_groebner_worker.py): the call now passes explicit `resource_limits=ProcessResourceLimits(cpu_seconds=10, address_space_bytes=1 GiB, file_size_bytes=1 MiB)` — mirroring the Singular adapter's caps — with `platform_tools=ProcessPlatformTools(prlimit_executable=...)` discovered in this registered owner module, so the child is hard-capped on CPU, address space, and output even before the wall-clock kill.

Validation: ruff check/format clean repo-wide, mypy strict clean, architecture gateway OK, lint-imports clean, tests/math/polynomials 97 passed. Pushed as 576fe99ec.

File: src/jacobian/math/polynomials/_models.py [hunk 10]
```diff
@@ -316,6 +327,307 @@ def require_canonical_basis_ring(self) -> Self:
...
+class IdealNormalFormResult(StrictModel):
+    """The canonical remainder of one polynomial modulo an ideal.
+
+    The result retains its complete source and the computed Gröbner basis so
+    validation replays the defining reduction relation exactly.  When the
+    exact remainder exceeds the 1,024-term output boundary the status reports
+    ``BUDGET_EXCEEDED`` instead of a mathematical conclusion; no partial
+    remainder is returned.  ``UNKNOWN`` reports that the bounded kernel could
+    not complete any strategy within its work bounds and asserts neither a
+    reduction nor a budget overflow; it carries no partial artifact.
```
[the remark this answers is not in the document]
Comment (morluto) [reply src/jacobian/math/polynomials/_models.py]: Fixed in 2dc3c1638: both published result schemas (IdealNormalFormResult and IdealMembershipResult) now document BUDGET_EXCEEDED as covering the two separately bounded exact artifacts — an oversized Gröbner-basis certificate (basis stripped) or an oversized canonical remainder / normal form (computed basis retained, no partial artifact) — so callers can distinguish a basis-certificate overflow from a remainder overflow. Your reproduced case (<x-(1+y+z)^4, x^12> queried with x^12, exact remainder zero) returns BUDGET_EXCEEDED with the basis stripped, i.e. an oversized certificate that makes no remainder claim; regression test test_budget_exceeded_semantics_cover_both_bounded_artifacts pins both schema descriptions and the existing test_complete_basis_output_overflow_reports_typed_outcome pins the behavior.

File: src/jacobian/math/polynomials/_operations.py [hunk 11]
```diff
@@ -215,3 +234,134 @@ def polynomial_groebner_basis(
...
+        return _incomplete_normal_form(request, status, None)
+    # Reduce with a bounded sparse-ring division so an admitted request can
+    # never expand an unbounded intermediate remainder before the 1,024-term
+    # output boundary is noticed; overflow becomes the typed budget outcome.
+    ring_context = _sparse_ring(variables, request.monomial_order)
+    divisors = [_ring_element(ring_context, element) for element in wire_basis]
+    dividend = _ring_element(ring_context, request.polynomial)
+    replayed = budgeted_reduce(ring_context, dividend, divisors)
+    if replayed is None:
+        return _incomplete_normal_form(request, "BUDGET_EXCEEDED", wire_basis)
```
[the remark this answers is not in the document]
Comment (morluto) [reply src/jacobian/math/polynomials/_operations.py:302]: Fixed in 02d01c12c: `budgeted_reduce` now returns `None` only when the exact remainder leaves the 1,024-term output boundary and raises `ReductionWorkLimitExceeded` when division steps or a temporarily expanded intermediate exhaust the work caps. The operation maps work exhaustion to the typed non-conclusion (UNKNOWN, no partial artifact), and the substantiation predicate no longer accepts work exhaustion as overflow evidence, so forged BUDGET_EXCEEDED claims for this shape are rejected by validation. Regression tests cover both the counterexample ideal and the forged-claim rejection; one established expectation that encoded the old conflation was updated to expect UNKNOWN.

File: src/jacobian/math/polynomials/_groebner_worker.py [hunk 12]
```diff
@@ -0,0 +1,235 @@
...
+                cpu_seconds=_CPU_SECONDS,
+                address_space_bytes=_ADDRESS_SPACE_BYTES,
+                file_size_bytes=_FILE_SIZE_BYTES,
+            ),
+            platform_tools=ProcessPlatformTools(prlimit_executable=prlimit),
+        )
+    except Exception:
+        return "unavailable", None
+    if completed.timed_out or completed.cancelled or completed.returncode != 0:
+        return "unavailable", None
```
[the remark this answers is not in the document]
Comment (morluto) [reply src/jacobian/math/polynomials/_groebner_worker.py]: Fixed in 396af783f: worker transport faults are now distinct typed failures instead of collapsing into 'unavailable' -> UNKNOWN. Launch failures (OSError) raise `GroebnerWorkerLaunchError`, cancellation raises `GroebnerWorkerCancelledError`, and a nonzero exit or malformed report without a timeout raises `GroebnerWorkerExecutionError` — none of them can be mistaken for work exhaustion, and a broken worker installation now surfaces as the execution fault it is. The hard wall-clock safety net still yields the typed non-conclusion (unavailable -> UNKNOWN) because a runaway attempt is mathematical hardness, not a deployment fault. TestWorkerFailurePreservation covers all four paths, including that public operations propagate the typed launch failure rather than returning UNKNOWN (tests/math/polynomials/test_ideal_membership.py passed).

File: src/jacobian/math/polynomials/_models.py [hunk 13]
```diff
@@ -316,6 +327,313 @@ def require_canonical_basis_ring(self) -> Self:
...
+) -> None:
+    """Check the retained basis spans the ideal and is its reduced basis.
+
+    One kernel replay substantiates authenticity: equality with the
+    recomputed reduced Gröbner basis subsumes membership of every
+    retained element in the source ideal.
+    """
+
+    if not generators_reduce_to_zero(ideal, groebner_basis, monomial_order):
+        raise ValueError("retained basis does not reduce every ideal generator to zero")
```
[the remark this answers is not in the document]
Comment (morluto) [reply src/jacobian/math/polynomials/_models.py]: Fixed in e92f82413.

Reproduced at 396af783f: your lex request (`x-∏(1+y_i)`, each `y_i²`, and the redundant relation `x^12-∏(1+12y_i)` with six `y` variables, queried with `x`) computed the correct answer — 7-element basis `{x-f, y_i²}`, 64-term remainder `f`, `NOT_IN_IDEAL` — but then raised a result-model `ValidationError` ("retained basis does not reduce every ideal generator to zero"), because `generators_reduce_to_zero` caught its `ReductionWorkLimitError`: replaying the redundant relation substitutes toward `∏(1+y_i)^12`, whose monomial box exhausts the 4,096-term intermediate work cap even though every exact artifact is inside its boundary.

Fix: `_require_authentic_retained_basis` now relies solely on `retained_basis_is_groebner`. Equality with the recomputed reduced Gröbner basis subsumes membership of every presentation generator in the certified ideal, so acceptance never depends on any presentation generator surviving its own capped service-side division; a redundant relation that merely exhausts replay *work* can no longer reject an independently authenticated result. `generators_reduce_to_zero` is removed with its only caller.

Regression tests: `test_redundant_generator_replay_work_limit_keeps_computed_result` reproduces your exact counterexample and round-trips both result models through `model_validate`; `test_computed_outcome_is_independent_of_redundant_relation_position` moves the redundant relation across the presentation (first, last, middle) and asserts identical `NOT_IN_IDEAL` outcomes with identical remainders. Validation: ruff check/format clean, mypy strict clean, tests/math + tests/catalog + tests/dispatch + tests/mcp + tests/tooling = 2391 passed.

File: src/jacobian/math/polynomials/_groebner_worker.py [hunk 14]
```diff
@@ -0,0 +1,280 @@
...
+            timeout_seconds=_TIMEOUT_SECONDS,
+            environment=worker_environment(),
+            stdout_limit=_STDOUT_LIMIT,
+            stderr_limit=_STDERR_LIMIT,
+            resource_limits=ProcessResourceLimits(
+                cpu_seconds=_CPU_SECONDS,
+                address_space_bytes=_ADDRESS_SPACE_BYTES,
+                file_size_bytes=_FILE_SIZE_BYTES,
+            ),
+            platform_tools=ProcessPlatformTools(prlimit_executable=prlimit),
```
[the remark this answers is not in the document]
Comment (morluto) [reply src/jacobian/math/polynomials/_groebner_worker.py:373]: Fixed in e92f82413.

Confirmed: with `prlimit_executable=None` (supported hosts without util-linux `prlimit`, e.g. macOS), `_resource_limited_command` skips wrapping and `_apply_resource_limits` no-ops without `resource.prlimit`, so the advertised 1 GiB address-space cap was never installed there.

Fix — portable isolation plus fail-closed: every worker payload now carries the CPU/address-space/file-size limits, and the child installs the identical POSIX rlimits itself via stdlib `resource.setrlimit` immediately after reading stdin and before importing pydantic/sympy, so the hard 1 GiB cap holds on hosts without `prlimit` too. The parent keeps prlimit wrapping where available (defense in depth), and when neither mechanism exists on the host it refuses to launch at all, raising the typed `GroebnerWorkerLaunchError` ("requires a hard memory limit...") instead of running an admitted request against unbounded memory; a child-side `setrlimit` failure likewise dies loudly rather than running unisolated.

Tests added under `TestPortableMemoryIsolation`: the payload carries the limits (and the parent still passes them to `run_bounded_process`); fail-closed when neither `prlimit` nor the resource module exists; and a direct child-level proof — launched bare without any prlimit wrapper, the worker dies nonzero under a 64 MiB address space yet reports `status ok` under the advertised 1 GiB cap. Validation: ruff check/format clean, mypy strict clean, tests/math + tests/catalog + tests/dispatch + tests/mcp + tests/tooling = 2391 passed.

File: src/jacobian/math/polynomials/_groebner_worker.py [hunk 15]
```diff
@@ -0,0 +1,280 @@
...
+        raise GroebnerWorkerLaunchError(str(error)) from None
+    if completed.cancelled:
+        raise GroebnerWorkerCancelledError(
+            "groebner worker was cancelled before reporting"
+        )
+    if completed.timed_out:
+        # The hard wall-clock safety net expired on a runaway attempt: the
+        # worker produced no report, so there is no conclusion either way
+        # about the ideal (a mathematically hard request).
+        return "unavailable", None
```
[the remark this answers is not in the document]
Comment (morluto) [reply src/jacobian/math/polynomials/_groebner_worker.py]: Fixed in e92f82413.

Agreed: a wall-clock expiry is deployment timing, not mathematics. Projecting it into UNKNOWN made the same request's answer depend on host load and concealed a retryable transport condition behind a domain status.

Fix: complete_basis_in_worker now returns a tagged ("timeout", None) attempt outcome instead of "unavailable". The orchestrator absorbs it only as far as retiring that one strategy — a later strategy may still conclude quickly, which keeps first-strategy flameouts from poisoning otherwise-computable requests — and when no strategy concluded, incremental_source_groebner raises the new typed GroebnerWorkerTimeoutError (a GroebnerWorkerError subclass) instead of letting either public operation report UNKNOWN. Both result schemas' UNKNOWN descriptions were updated to state that it means every bounded strategy concluded without evidence inside its work bounds, explicitly excluding wall-clock expiry.

Tests: test_worker_timeout_is_a_typed_transport_failure (attempt tag, orchestrator raise), test_timeout_does_not_conceal_a_later_concluding_strategy (timeout then ok still yields the basis), test_worker_timeout_raises_typed_failure_from_operations (both public operations raise the typed failure); the eight-variable C-chain regression now asserts GroebnerWorkerTimeoutError from both operations instead of UNKNOWN. Validation: ruff check/format clean, mypy strict clean, tests/math + tests/catalog + tests/dispatch + tests/mcp + tests/tooling = 2391 passed.

File: src/jacobian/math/polynomials/_models.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (morluto) [reply src/jacobian/math/polynomials/_models.py]: Fixed in bbbdd7384: `generators_reduce_to_zero` now returns a tri-state verdict — True, False (provably nonzero remainder), or None when the capped replay exhausted its work budget on some presentation generator. The result validator treats None as inconclusive and still accepts on the independent reduced-basis recomputation, so the six-variable lex example with a redundant exploding generator no longer turns a successfully computed answer into a result-model ValidationError.

File: src/jacobian/math/polynomials/_groebner_worker.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (morluto) [reply src/jacobian/math/polynomials/_groebner_worker.py:373]: Fixed in bbbdd7384: worker launch now fails closed (`GroebnerWorkerLaunchError: hard address-space cap cannot be enforced...`) when neither util-linux prlimit nor resource.prlimit is available, instead of silently running without the advertised 1 GiB cap. test_memory_cap_enforcement_fails_closed_without_prlimit covers both platform branches.

File: src/jacobian/math/polynomials/_groebner_worker.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (morluto) [reply src/jacobian/math/polynomials/_groebner_worker.py]: Fixed in bbbdd7384: wall-clock timeouts raise the new typed `GroebnerWorkerTimeoutError` (a retryable transport failure) rather than collapsing to 'unavailable' -> UNKNOWN, matching cancellation and launch-failure handling; WorkerStatus shrinks to kernel-reported outcomes only. The previously timeout-dependent UNKNOWN test was updated to expect the typed error, and TestWorkerFailurePreservation::test_hard_timeout_raises_typed_transport_failure pins it.

File: src/jacobian/math/polynomials/_groebner_worker.py [hunk 16]
```diff
@@ -0,0 +1,359 @@
...
+) -> tuple[WorkerStatus, tuple[RationalPolynomial, ...] | None]:
+    """Classify the tagged report of a cleanly completed worker run."""
+
+    from pydantic import ValidationError
+
+    try:
+        report = json.loads(completed.stdout.decode("utf-8"))
+    except Exception as error:
+        raise GroebnerWorkerExecutionError(str(error)) from None
+    status = report.get("status")
```
[the remark this answers is not in the document]
Comment (morluto) [reply src/jacobian/math/polynomials/_groebner_worker.py]: Root cause: `complete_basis_in_worker` decoded the worker's stdout with `json.loads` and immediately called `report.get("status")`, so any valid-JSON payload with a non-object top level (`[]`, `null`, a string, a number) escaped as an untyped `AttributeError`.

Fix: added `_decode_report` in src/jacobian/math/polynomials/_groebner_worker.py, which fails closed with `GroebnerWorkerExecutionError("groebner worker report is <type>, not a JSON object")` when the decoded report is not a mapping, before any field access.

Validation: reproduced first (worker report `[]`/`null`/`"ok"`/`42` each raised `AttributeError` before the fix); after the fix all four raise `GroebnerWorkerExecutionError`. New tests: `TestWorkerFailurePreservation.test_non_object_report_raises_execution_error` (parametrized over those payloads). Lane `tests/math/polynomials`: 112 passed; ruff check/format and strict mypy clean.

File: src/jacobian/math/polynomials/_groebner_worker.py [hunk 17]
```diff
@@ -0,0 +1,304 @@
...
+            f"groebner worker exited with returncode "
+            f"{completed.returncode} without a report"
+        )
+    try:
+        report = json.loads(completed.stdout.decode("utf-8"))
+    except Exception as error:
+        raise GroebnerWorkerExecutionError(str(error)) from None
+    status = report.get("status")
+    if status in ("exceeded", "aborted", "failed"):
+        return status, None
```
[the remark this answers is not in the document]
Comment (morluto) [reply src/jacobian/math/polynomials/_groebner_worker.py]: Root cause: the parent treated the worker's self-reported `failed` status (emitted by `_WORKER_PROGRAM` when a codec or result-conversion exception fires) as an inconclusive strategy: `incremental_source_groebner` retried past it, and after three failures returned `(None, False)`, so both public operations reported UNKNOWN — a broken worker/backend adapter masquerading as mathematical work exhaustion. The return also violated `WorkerStatus`, which never included `"failed"`.

Fix: in src/jacobian/math/polynomials/_groebner_worker.py, `complete_basis_in_worker` now raises `GroebnerWorkerExecutionError` for a `failed` report ("internal kernel or codec failure without any mathematical conclusion"); only `aborted` remains a bounded non-conclusion and `exceeded` an evidenced overflow. Docstrings of the worker module and `incremental_source_groebner` updated accordingly.

Validation: reproduced first — patching the worker to emit `{"status": "failed"}` made both operations return UNKNOWN; they now raise `GroebnerWorkerExecutionError` while the real-worker path still returns COMPUTED for a known answer. New tests: `test_failed_worker_report_raises_execution_error` and `test_public_operations_surface_failed_worker_reports`. Lane `tests/math/polynomials`: 112 passed; ruff check/format and strict mypy clean.

File: src/jacobian/math/polynomials/_groebner_worker.py [hunk 18]
```diff
@@ -0,0 +1,334 @@
...
+
+    if strategy == "complete":
+        try:
+            basis = groebner(
+                generators, *symbols,
+                order=payload["monomial_order"], domain=QQ,
+            )
+        except Exception:
+            print(json.dumps({"status": "failed"}))
+            return
```
[the remark this answers is not in the document]
Comment (morluto) [reply src/jacobian/math/polynomials/_groebner_worker.py]: Fixed in 08613d25c: the worker now separates MemoryError from other kernel exceptions inside one attempt boundary (`_kernel_attempt`), covering the kernel call, basis materialization into `Poly`, the output-budget scan, and wire conversion/serialization in `_emit_ok`. It emits a typed `"exhausted"` report from a pre-serialized literal so the emission itself cannot need a large allocation right after an allocation failure. The parent accepts `"exhausted"` as a bounded non-conclusion status and `_replay.incremental_source_groebner` falls through to the next strategy, ending in UNKNOWN when no strategy concludes — no more GroebnerWorkerExecutionError for memory-capped admitted work.

File: src/jacobian/math/polynomials/_groebner_worker.py [hunk 19]
```diff
@@ -0,0 +1,334 @@
...
+
+def _basis_from_report(report: dict[str, Any]) -> tuple[RationalPolynomial, ...]:
+    """Convert a worker ``ok`` report to the canonical wire basis."""
+
+    from pydantic import ValidationError
+
+    try:
+        return tuple(
+            RationalPolynomial.model_validate(element) for element in report["basis"]
+        )
```
[the remark this answers is not in the document]
Comment (morluto) [reply src/jacobian/math/polynomials/_groebner_worker.py:290]: Fixed in 08613d25c: `_basis_from_report` now validates the complete returned-basis contract at the worker protocol layer before returning any `ok` payload — every element must use the submitted ideal's exact ordered ring, no zero elements, and aggregate term count within the 1,024-term output boundary — raising `GroebnerWorkerExecutionError` otherwise, so a malformed nested response can no longer escape as a generic result-model ValidationError or drive in-process reduction over foreign-ring data. Regression tests cover foreign/reordered rings, zero elements, aggregate overflow, and the public-operation typed fault.

**Comment by morluto** [comment]
Closing as superseded: #2240 landed the Gröbner family on main (issue #1685) with `polynomial.ideal.normal_form.compute` returning the exact remainder plus the `in_ideal` membership boolean — covering both postconditions this branch added (`membership.decide`, `normal_form.compute`, same ID). Keeping both would duplicate catalog IDs (the source of the current `operation admission IDs must be unique` CI failures).

Issue #1665 stays open for its remaining distinct scope (quotient ring reduction), which neither this branch nor #2240 publishes.
